Add locked metric and missing connection id

pudkrong opened the pull request:
During our load test with the default configuration (where shadow fetching occurs every 5 seconds), we observed a significant number of missing shadow messages compared to our initial expectations. To investigate this issue, we implemented addition metrics to gain better insights into the problem:

### FetchDeviceShadow Function

1.  `locked`: This metric counts instances when the function is still processing a previous request when it is invoked again.

### OnMessage Function

During the load test, we encountered errors in the `onMessage` function, preventing it from updating the database as expected when certain conditions were not met. To address this issue, we took the following actions:

1.  **Error Handling**: We introduced error handling mechanisms to properly manage errors encountered in the `onMessage` function.
    
2.  `connectionIdMissing`: This metric was added to count the number of times the `onMessage` function failed to update the data due to condition mismatches.

### FetchDeviceShadow invocation rate

During our load test, we expected the `fetchDeviceShadow` function to be invoked 12 times per minute, as it is triggered by SQS every 5 seconds. However, we observed that the average invocation rate was only 6 times per minute. After a thorough investigation, we determined that this discrepancy was due to a concurrency limit of 2 on the function. To align with our desired invocation rate, we made the necessary adjustment by increasing the `maxConcurrency` to 15.

<img width="1317" alt="MicrosoftTeams-imagea9e62581afd4e14826cb00a487492cc5140946d32fabde4e2c44d75bd88198d4" src="https://github.com/hello-nrfcloud/backend/assets/2525169/7420ff6c-ed78-4f9f-90b2-0c66055deb56">

coderbyheart reviewed (changes requested):
- has no tests, so unclear why this is needed
- commit has no explanation why this is needed

Fix at least one of those.

coderbyheart reviewed (changes requested):
Still no test or explanation why change is needed in commit message.

pudkrong committed:
docs: add locked metric and missing connection id

During our load test with the default configuration (where shadow fetching occurs every 5 seconds), we observed a significant number of missing shadow messages compared to our initial expectations. To investigate this issue, the addition metrics are added to gain better insights

- locked: This metric counts instances when the function is still processing a previous request when it is invoked again.
- connectionIdMissing: This metric was added to count the number of times the onMessage function failed to update the data due to condition mismatches.

pudkrong committed:
fix: increase concurrency rate

We expected the fetchDeviceShadow function to be invoked 12 times per minute, as it is triggered by SQS every 5 seconds. However, during the load test we observed that the average invocation rate was only 6 times per minute. After a thorough investigation, we determined that this discrepancy was due to a concurrency limit of 2 on the function. To align with our desired invocation rate, we increase the maxConcurrency to 15.

coderbyheart commented on websocket/deviceShadowUpdateChecker.ts:
```diff
@@ -100,7 +100,7 @@ export const createDeviceUpdateChecker = async (
 			{ count: Number.MAX_SAFE_INTEGER, interval: 5 },
 		]
-		const deviceModelHash = hashSHA1(device.model)
+		const deviceModelHash = hashSHA1(device.model ?? '')
```
```suggestion
		const deviceModelHash = hashSHA1(device.model)
```

This should never happen. If a device has no model it is badly configured, do not fetch shadow for these devices.

pudkrong replied:
I do agree that it should not happen, but I found this case during the load test. This happened before I add `attribute_exists(#connectionId)` into the condition expression per this https://github.com/hello-nrfcloud/backend/blob/c1bfc89e39a270e86f7ce708d301f85b62453a03/websocket/connectionsRepository.ts#L143.

Without having the connection Id in condition expression, it just adds data as a new item without model. 

Even though I have fixed this, but just in case the model does not exist, it will throw the error because hash function requires string, not undefined.

coderbyheart replied:
But then it can be reverted now, because this will no longer happen.

coderbyheart commented on lambda/onMessage.ts:
```diff
@@ -95,7 +98,16 @@ const success =
+			log.debug(`WebConnection is not found`, {
+				context: event.requestContext.authorizer,
+			})
+			track('ConnectionIdMissing', MetricUnits.Count, 1)
```
*problem:* it is not checked why `extendTTL` fails. It may not be cause by a missing connectionId.

pudkrong replied:
This metric will be logged only when the error is `ConditionalCheckFailedException` as per https://github.com/hello-nrfcloud/backend/pull/234/files/c1bfc89e39a270e86f7ce708d301f85b62453a03#diff-900a11ad083b712ce3f9a5b3e1df19c6723e1edae68a0be3d0d9a38252cf2ec4R104.

If error is something else, it will just ignore.

coderbyheart reviewed (changes requested):
Why is the problem not fixed that was identified (missing connectionId)?

pudkrong commented:
> Why is the problem not fixed that was identified (missing connectionId)?

Actually, it should never happen if we keep sending ping within 5 minutes (DynamoDB TTL). But the data tells it still happens but is not significant, less than 1% according to the graph

<img width="1338" alt="SCR-25660907-pgdj" src="https://github.com/hello-nrfcloud/backend/assets/2525169/e333670d-610c-4034-b772-cf9f799a18cd">

There are only 2 events that delete the connection data
- Web socket disconnects  (immediately delete)
- Web socket was lost and cannot keep alive for more than 5 minutes

To be honest, I still don't know why it happens but I would like to keep the metric for now. So, in the future if the number increases significant, I can spend more time to investigate this.

coderbyheart reviewed (changes requested):
Thanks. 

Please squash into two commits, and change them to 'fix' because they do not add new features (that are relevant to the user).

Arguably the metric commit could be a `doc` commit, since it does not even change the behavior of the software.
